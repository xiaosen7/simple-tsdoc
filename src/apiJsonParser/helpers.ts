import * as model from "@microsoft/api-extractor-model";
import * as types from "../types";
import * as tsdoc from "@microsoft/tsdoc";
import * as commentParser from "../commentParser";

const canVisitKinds = [
  model.ApiItemKind.Class,
  model.ApiItemKind.Function,
  model.ApiItemKind.Method,
  model.ApiItemKind.Property,
  model.ApiItemKind.Constructor,
  model.ApiItemKind.Namespace,
] as const;
type CanVisitKind = typeof canVisitKinds[number];
type AfterVisitKind = `After${CanVisitKind}`;
export type Visitor = Partial<
  Record<CanVisitKind | AfterVisitKind, (apiDocItem: types.ApiDocItem) => any>
>;

export function getAnnotation(apiItem: model.ApiDeclaredItem) {
  const tsdocComment: tsdoc.DocComment | undefined = apiItem.tsdocComment;

  let annotation = defaultAnnotation;
  if (tsdocComment) {
    annotation = commentParser.parse(tsdocComment.emitAsTsdoc());
  }

  return annotation;
}

export function traversePackage(
  apiPackage: model.ApiPackage,
  visitor: Visitor
) {
  traverseMembers(apiPackage.members, visitor);
}

function traverseCallback(apiItem: model.ApiItem, visitor: Visitor) {
  let apiDocItem!: types.ApiDocItem;
  let kind!: model.ApiItemKind;
  let canVisit = false;

  if (apiItem instanceof model.ApiDeclaredItem) {
    const annotation = getAnnotation(apiItem);
    kind = getKind(apiItem, annotation);

    if (canVisitKinds.includes(kind as any)) {
      apiDocItem = {
        annotation,
        kind,
        name: apiItem.displayName,
        apiItem,
      };

      canVisit = true;
    }
  }

  if (canVisit) {
    const fn = visitor[kind as CanVisitKind];
    fn && fn(apiDocItem);
  }

  if (apiItem.members) {
    traverseMembers(apiItem.members, visitor);
  }

  if (canVisit) {
    const after = visitor[getAfterKind(kind as CanVisitKind)];
    after && after(apiDocItem);
  }
}

function traverseMembers(members: readonly model.ApiItem[], visitor: Visitor) {
  if (members?.length) {
    members.forEach((apiItem) => traverseCallback(apiItem, visitor));
  }
}
export const defaultAnnotation: types.Annotation = {
  tags: [],
  deprecated: false,
};

export function isFunctionKind(
  apiItem: model.ApiItem,
  annotation: types.Annotation
) {
  return (
    apiItem.kind === model.ApiItemKind.Function ||
    (apiItem.kind === model.ApiItemKind.Variable &&
      annotation.modifierTags?.includes("@function"))
  );
}

export function isClassKind(
  apiItem: model.ApiItem,
  annotation: types.Annotation
) {
  return (
    apiItem.kind === model.ApiItemKind.Class ||
    (apiItem.kind === model.ApiItemKind.Variable &&
      annotation.modifierTags?.includes("@constructor"))
  );
}

export function getKind(
  apiItem: model.ApiItem,
  annotation: types.Annotation
): model.ApiItemKind {
  if (isFunctionKind(apiItem, annotation)) {
    return model.ApiItemKind.Function;
  }

  if (isClassKind(apiItem, annotation)) {
    return model.ApiItemKind.Class;
  }

  return apiItem.kind;
}

export function getAfterKind(kind: CanVisitKind): AfterVisitKind {
  return `After${kind}` as const;
}
