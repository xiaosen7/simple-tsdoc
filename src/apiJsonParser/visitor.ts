import * as helpers from "./helpers";
import * as model from "@microsoft/api-extractor-model";
import { ApiDocItem } from "../types";
import * as utils from "../utils";

export function createVisitor(docItems: ApiDocItem[]): helpers.Visitor {
  const classApiDocItemStack: ApiDocItem[] = [];
  const namespaceApiDocItemStack: ApiDocItem[] = [];

  return {
    [model.ApiItemKind.Function](apiDocItem) {
      if (utils.peek(namespaceApiDocItemStack)) {
        pushIntoProperties(namespaceApiDocItemStack, apiDocItem);
      } else {
        docItems.push(apiDocItem);
      }
    },
    [model.ApiItemKind.Class](apiDocItem) {
      classApiDocItemStack.push(apiDocItem);
      if (utils.peek(namespaceApiDocItemStack)) {
        pushIntoProperties(namespaceApiDocItemStack, apiDocItem);
      } else {
        docItems.push(apiDocItem);
      }
    },
    [helpers.getAfterKind(model.ApiItemKind.Class)]() {
      classApiDocItemStack.pop();
    },
    [model.ApiItemKind.Property](apiDocItem) {
      pushIntoProperties(classApiDocItemStack, apiDocItem);
    },
    [model.ApiItemKind.Method](apiDocItem) {
      pushIntoProperties(classApiDocItemStack, apiDocItem);
    },
    [model.ApiItemKind.Constructor](apiDocItem) {
      pushIntoProperties(classApiDocItemStack, apiDocItem);
    },
    [model.ApiItemKind.Namespace](apiDocItem) {
      namespaceApiDocItemStack.push(apiDocItem);
      docItems.push(apiDocItem);
    },
    [helpers.getAfterKind(model.ApiItemKind.Namespace)]() {
      namespaceApiDocItemStack.pop();
    },
  };
}

function pushIntoProperties(parentStack: ApiDocItem[], apiDocItem: ApiDocItem) {
  const parent = utils.peek(parentStack);
  if (parent) {
    if (!parent.properties) {
      parent.properties = [];
    }

    parent.properties.push(apiDocItem);
  }
}
