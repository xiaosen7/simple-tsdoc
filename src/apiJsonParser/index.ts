import * as model from "@microsoft/api-extractor-model";
import { ApiDocItem } from "../types";
import { traversePackage } from "./helpers";
import { createVisitor } from "./visitor";

export function parse(apiJsonFilePath: string) {
  const apiModel: model.ApiModel = new model.ApiModel();
  const apiPackage: model.ApiPackage = apiModel.loadPackage(apiJsonFilePath);

  const docItems: ApiDocItem[] = [];
  const visitor = createVisitor(docItems);
  traversePackage(apiPackage, visitor);

  return docItems;
}
