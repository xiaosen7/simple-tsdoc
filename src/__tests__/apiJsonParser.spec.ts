import * as apiJsonParser from "../apiJsonParser";
import * as testUtils from "./testUtils";

test("apiJsonParser", () => {
  const inputFilePath = testUtils.resolveAssetPath(
    "api-extractor-demo.api.json"
  );
  const docItems = apiJsonParser.parse(inputFilePath);
  expect(docItems.map(({ apiItem, ...rest }) => rest)).toMatchSnapshot();
});
