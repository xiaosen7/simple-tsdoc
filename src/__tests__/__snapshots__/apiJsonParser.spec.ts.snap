// Vitest Snapshot v1

exports[`apiJsonParser 1`] = `
[
  {
    "annotation": {
      "deprecated": false,
      "description": "add 函数",
      "modifierTags": [
        "@public",
      ],
      "namespace": "math",
      "params": [],
      "returns": undefined,
      "tags": [],
    },
    "kind": "Function",
    "name": "add",
  },
  {
    "annotation": {
      "deprecated": false,
      "description": "",
      "modifierTags": [
        "@function",
      ],
      "namespace": "其他",
      "params": [],
      "returns": undefined,
      "tags": [],
    },
    "kind": "Function",
    "name": "arrowFunction",
  },
  {
    "annotation": {
      "deprecated": false,
      "description": "",
      "modifierTags": [
        "@constructor",
      ],
      "namespace": "其他",
      "params": [],
      "returns": undefined,
      "tags": [],
    },
    "kind": "Class",
    "name": "createdClass",
    "properties": [],
  },
  {
    "annotation": {
      "deprecated": false,
      "description": "",
      "modifierTags": [
        "@function",
      ],
      "namespace": "其他",
      "params": [],
      "returns": undefined,
      "tags": [],
    },
    "kind": "Function",
    "name": "createdFunction",
  },
  {
    "annotation": {
      "deprecated": false,
      "description": "foo函数",
      "modifierTags": [],
      "namespace": "foobar",
      "params": [
        {
          "description": "",
          "name": "name",
        },
        {
          "description": "",
          "name": "age",
        },
      ],
      "returns": "",
      "tags": [
        {
          "description": "\`\`\`ts
const res = foo('xiaoming', 18); // => xiaoming18
\`\`\`",
          "name": "@example",
        },
      ],
    },
    "kind": "Function",
    "name": "foo",
  },
  {
    "annotation": {
      "deprecated": false,
      "description": "人这个类",
      "modifierTags": [],
      "namespace": "其他",
      "params": [],
      "returns": undefined,
      "tags": [],
    },
    "kind": "Class",
    "name": "Person",
    "properties": [
      {
        "annotation": {
          "deprecated": false,
          "description": "Constructs a new instance of the \`Person\` class",
          "modifierTags": [],
          "namespace": "Person",
          "params": [],
          "returns": undefined,
          "tags": [],
        },
        "kind": "Constructor",
        "name": "(constructor)",
      },
      {
        "annotation": {
          "deprecated": false,
          "namespace": "Person",
          "tags": [],
        },
        "kind": "Method",
        "name": "getName",
      },
      {
        "annotation": {
          "deprecated": false,
          "namespace": "Person",
          "tags": [],
        },
        "kind": "Property",
        "name": "name",
      },
    ],
  },
]
`;
